### ISSUE TITLE
Habitat version 1.5.0 runs into errors with proxies.

### ISSUE THREAD

**Issue by mike10010100** [habitat-sh/habitat#7448]
OS: Ubuntu 16.04
Debug: not yet available
Hab environment variables:
- HTTP_PROXY
- HTTPS_PROXY
- http_proxy
- https_proxy
  - All http_proxy/https_proxy point to the same username/password auth'd http proxy endpoint with custom port: http://<username>:<password>@proxy.url.com:<port>
- HAB_BLDR_URL=https://<private_depot_url>
When these variables are set on habitat or the habitat supervisor (version 1.5.0), and `hab pkg install` is run for any package, the following error occurs:

```error sending request for url (https://<depot_url>/v1/depot/channels/<origin>/stable/pkgs/<pkg>/latest?target=x86_64-linux): error trying to connect: dns error: failed to lookup address information: Name does not resolve```

The behavior occurs regardless of whether or not the BLDR_URL or HAB_AUTH_TOKEN are set.

**Comment by fifthecho** [comment]
It is worth noting that this is an environment where the DNS for the depot does not resolve from the host operating system, but through the HTTP(S) proxy.

**Comment by christophermaier** [comment]
My initial suspicion is that this may be an unintended side effect of our recent migration of the HTTP client backend.

**Comment by stevendanna** [comment]
TLDR, I think this is a version compatibility problem between some of our crates that are causing us to ignore proxy configuration. Details here: https://github.com/inejge/env_proxy/issues/8

Debugging notes below:

In my local testing, it appears that HTTPS_PROXY is not being respected at all.

On 0.90.6, if I set HTTPS_PROXY, requests go through the proxy as expected.  The following is the RUST_LOG=debug output from such a run:

```
# RUST_LOG=debug SSL_CERT_FILE=~/.mitmproxy/mitmproxy-ca-cert.pem HTTPS_PROXY=http://localhost:8080 hab pkg exec core/hab/0.90.6 hab pkg search core/glibc
[2020-03-04T10:19:49Z DEBUG habitat_common::ui] UI { shell: Shell { input: InputStream { isatty: true }, out: OutputStream { coloring: Auto, isatty: true }, err: OutputStream { coloring: Auto, isatty: true } } }
[2020-03-04T10:19:49Z DEBUG hab] clap cli args: ["hab", "pkg", "exec", "core/hab/0.90.6", "hab"]
[2020-03-04T10:19:49Z DEBUG hab] remaining cli args: ["pkg", "search", "core/glibc"]
[2020-03-04T10:19:49Z DEBUG hab::command::pkg::exec] Setting: PATH='/hab/pkgs/core/hab/0.90.6/20191112141314/bin'
[2020-03-04T10:19:49Z DEBUG hab::command::pkg::exec] Running: /hab/pkgs/core/hab/0.90.6/20191112141314/bin/hab pkg search core/glibc
[2020-03-04T10:19:49Z DEBUG habitat_core::os::process::unix] Calling execvp(): ("/hab/pkgs/core/hab/0.90.6/20191112141314/bin/hab") ["pkg", "search", "core/glibc"]
[2020-03-04T10:19:49Z DEBUG habitat_common::ui] UI { shell: Shell { input: InputStream { isatty: true }, out: OutputStream { coloring: Auto, isatty: true }, err: OutputStream { coloring: Auto, isatty: true } } }
[2020-03-04T10:19:49Z DEBUG hab] clap cli args: ["/hab/pkgs/core/hab/0.90.6/20191112141314/bin/hab", "pkg", "search", "core/glibc"]
[2020-03-04T10:19:49Z DEBUG hab] remaining cli args: []
[2020-03-04T10:19:49Z DEBUG hab::config] No CLI config found, loading defaults
[2020-03-04T10:19:49Z DEBUG habitat_http_client::api_client] Client socket timeout: 300 secs
[2020-03-04T10:19:49Z DEBUG habitat_http_client::api_client] Skip cert verification: false
[2020-03-04T10:19:49Z DEBUG habitat_http_client::api_client] User-Agent: hab/0.90.6/20191112141314 (x86_64-linux; 4.19.76-linuxkit)
[2020-03-04T10:19:49Z DEBUG habitat_http_client::api_client] Setting https proxy to http://localhost:8080/
[2020-03-04T10:19:49Z DEBUG habitat_http_client::api_client] Found an installed Habitat core/cacerts package at: /hab/pkgs/core/cacerts/2018.12.05/20190115014206/ssl/cert.pem
[2020-03-04T10:19:49Z DEBUG habitat_http_client::api_client] Processing cert file: /hab/pkgs/core/cacerts/2018.12.05/20190115014206/ssl/cert.pem
[2020-03-04T10:19:49Z DEBUG habitat_http_client::api_client] Processing cert file: /hab/cache/ssl/cert.pem
[2020-03-04T10:19:49Z DEBUG habitat_api_client::builder] Searching for package core/glibc with range 0
[2020-03-04T10:19:49Z DEBUG habitat_http_client::api_client] GET https://bldr.habitat.sh/v1/depot/pkgs/search/core%2Fglibc?range=0&distinct=true with ApiClient { endpoint: "https://bldr.habitat.sh/v1", inner: Client }
[2020-03-04T10:19:49Z DEBUG hyper::client::connect::dns] resolving host="localhost"
[2020-03-04T10:19:49Z DEBUG hyper::client::connect::http] connecting to 127.0.0.1:8080
[2020-03-04T10:19:49Z DEBUG tokio_reactor] adding I/O source: 0
[2020-03-04T10:19:49Z DEBUG tokio_reactor::registration] scheduling Write for: 0
[2020-03-04T10:19:49Z DEBUG hyper::client::connect::http] connected to Some(V4(127.0.0.1:8080))
[2020-03-04T10:19:49Z DEBUG tokio_reactor::registration] scheduling Read for: 0
[2020-03-04T10:19:49Z DEBUG tokio_reactor::registration] scheduling Read for: 0
[2020-03-04T10:19:49Z DEBUG tokio_reactor::registration] scheduling Read for: 0
[2020-03-04T10:19:49Z DEBUG tokio_reactor::registration] scheduling Read for: 0
[2020-03-04T10:19:49Z DEBUG tokio_reactor::registration] scheduling Read for: 0
[2020-03-04T10:19:49Z DEBUG tokio_reactor::registration] scheduling Read for: 0
[2020-03-04T10:19:49Z DEBUG hyper::proto::h1::io] flushed 420 bytes
[2020-03-04T10:19:49Z DEBUG hyper::proto::h1::io] read 671 bytes
[2020-03-04T10:19:49Z DEBUG hyper::proto::h1::io] parsed 24 headers
[2020-03-04T10:19:49Z DEBUG hyper::proto::h1::conn] incoming body is content-length (119 bytes)
[2020-03-04T10:19:49Z DEBUG tokio_reactor::registration] scheduling Read for: 0
[2020-03-04T10:19:49Z DEBUG reqwest::async_impl::response] Response: '200 OK' for https://bldr.habitat.sh/v1/depot/pkgs/search/core%2Fglibc?range=0&distinct=true
[2020-03-04T10:19:49Z DEBUG tokio_reactor::registration] scheduling Read for: 0
[2020-03-04T10:19:49Z DEBUG habitat_api_client::builder] Response Status: 200
[2020-03-04T10:19:49Z DEBUG hyper::proto::h1::io] read 119 bytes
[2020-03-04T10:19:49Z DEBUG hyper::proto::h1::conn] incoming body completed
[2020-03-04T10:19:49Z DEBUG tokio_reactor] dropping I/O source: 0
```

However, on 1.5.0 and 1.5.50, the request does not hit the proxy:

```
# RUST_LOG=debug SSL_CERT_FILE=~/.mitmproxy/mitmproxy-ca-cert.pem HTTPS_PROXY=http://localhost:8080 hab pkg exec core/hab/1.5.0 hab pkg search core/glibc
[2020-03-04T10:20:02Z DEBUG habitat_common::ui] UI { shell: Shell { input: InputStream { isatty: true }, out: OutputStream { coloring: Auto, isatty: true }, err: OutputStream { coloring: Auto, isatty: true } } }
[2020-03-04T10:20:02Z DEBUG hab] clap cli args: ["hab", "pkg", "exec", "core/hab/1.5.0", "hab"]
[2020-03-04T10:20:02Z DEBUG hab] remaining cli args: ["pkg", "search", "core/glibc"]
[2020-03-04T10:20:02Z DEBUG hab::command::pkg::exec] Setting: PATH='/hab/pkgs/core/hab/1.5.0/20200122222330/bin'
[2020-03-04T10:20:02Z DEBUG hab::command::pkg::exec] Running: /hab/pkgs/core/hab/1.5.0/20200122222330/bin/hab pkg search core/glibc
[2020-03-04T10:20:02Z DEBUG habitat_core::os::process::unix] Calling execvp(): ("/hab/pkgs/core/hab/1.5.0/20200122222330/bin/hab") ["pkg", "search", "core/glibc"]
[2020-03-04T10:20:02Z DEBUG habitat_common::ui] UI { shell: Shell { input: InputStream { isatty: true }, out: OutputStream { coloring: Auto, isatty: true }, err: OutputStream { coloring: Auto, isatty: true } } }
[2020-03-04T10:20:02Z DEBUG hab] clap cli args: ["/hab/pkgs/core/hab/1.5.0/20200122222330/bin/hab", "pkg", "search", "core/glibc"]
[2020-03-04T10:20:02Z DEBUG hab] remaining cli args: []
[2020-03-04T10:20:02Z DEBUG hab::config] No CLI config found, loading defaults
[2020-03-04T10:20:02Z DEBUG habitat_http_client::api_client] Client socket timeout: 300 secs
[2020-03-04T10:20:02Z DEBUG habitat_http_client::api_client] Skip cert verification: false
[2020-03-04T10:20:02Z DEBUG habitat_http_client::api_client] User-Agent: hab/1.5.0/20200122222330 (x86_64-linux; 4.19.76-linuxkit)
[2020-03-04T10:20:02Z WARN  env_proxy] could not set URL scheme back to http
[2020-03-04T10:20:02Z DEBUG habitat_http_client::api_client] No proxy configured for url: "https://bldr.habitat.sh/v1"
[2020-03-04T10:20:02Z DEBUG habitat_http_client::api_client] Found an installed Habitat core/cacerts package at: /hab/pkgs/core/cacerts/2018.12.05/20190115014206/ssl/cert.pem
[2020-03-04T10:20:02Z DEBUG habitat_http_client::api_client] Processing cert file: /hab/pkgs/core/cacerts/2018.12.05/20190115014206/ssl/cert.pem
[2020-03-04T10:20:02Z DEBUG habitat_http_client::api_client] Processing cert file: /hab/cache/ssl/cert.pem
[2020-03-04T10:20:02Z DEBUG habitat_api_client::builder] Searching for package core/glibc with range 0
[2020-03-04T10:20:02Z DEBUG habitat_http_client::api_client] GET https://bldr.habitat.sh/v1/depot/pkgs/search/core%2Fglibc?range=0&distinct=true with ApiClient { endpoint: "https://bldr.habitat.sh/v1", inner: Client }
[2020-03-04T10:20:02Z DEBUG hyper::client::connect::dns] resolving host="bldr.habitat.sh"
[2020-03-04T10:20:02Z DEBUG hyper::client::connect::http] connecting to 151.101.17.209:443
[2020-03-04T10:20:02Z DEBUG hyper::client::connect::http] connected to Some(V4(151.101.17.209:443))
[2020-03-04T10:20:02Z DEBUG hyper::proto::h1::io] flushed 396 bytes
[2020-03-04T10:20:03Z DEBUG hyper::proto::h1::io] read 671 bytes
[2020-03-04T10:20:03Z DEBUG hyper::proto::h1::io] parsed 24 headers
[2020-03-04T10:20:03Z DEBUG hyper::proto::h1::conn] incoming body is content-length (119 bytes)
[2020-03-04T10:20:03Z DEBUG hyper::proto::h1::io] read 119 bytes
[2020-03-04T10:20:03Z DEBUG hyper::proto::h1::conn] incoming body completed
[2020-03-04T10:20:03Z DEBUG reqwest::async_impl::response] Response: '200 OK' for https://bldr.habitat.sh/v1/depot/pkgs/search/core%2Fglibc?range=0&distinct=true
[2020-03-04T10:20:03Z DEBUG habitat_api_client::builder] Response Status: 200
core/glibc
```

Note, in the working case (0.90.6) we see:

```
[2020-03-04T10:19:49Z DEBUG habitat_http_client::api_client] Setting https proxy to http://localhost:8080/
```

But in the non-working case (1.5.0+) we see:

```
[2020-03-04T10:20:02Z WARN  env_proxy] could not set URL scheme back to http
[2020-03-04T10:20:02Z DEBUG habitat_http_client::api_client] No proxy configured for url: "https://bldr.habitat.sh/v1"
```

I see the same behavior when I use no scheme, `http://`, or `https://`.

A small test program confirms that the problem is in the construction of the Url in the env_proxy library:

```
# cat src/main.rs
extern crate log;
extern crate env_proxy;

use url::Url;

fn main() {
    env_logger::init();
    let mut url = Url::parse("https://www.example.org").unwrap();
    let proxy = env_proxy::for_url(&url);
    println!("proxy.raw_value(): {:?}", proxy.raw_value());
    //println!("proxy.to_string(): {:?}", proxy.to_string());
}
# RUST_LOG=debug HTTPS_PROXY="https://localhost:3000" target/debug/env-proxy-tes
t                                                                                                                                                                                     
proxy.raw_value(): Some("https://localhost:3000")
# edit test program to print string                                                                                                                                   
# cargo build                                                                   
   Compiling env-proxy-test v0.1.0 (/src/env-proxy-test)                                                                                                                              
    Finished dev [unoptimized + debuginfo] target(s) in 21.90s                                                                                                                        
# RUST_LOG=debug HTTPS_PROXY="https://localhost:3000" target/debug/env-proxy-tes
t                                                                                                                                                                                     
[2020-03-04T10:55:55Z WARN  env_proxy] could not set URL scheme back to https                                                                                                         
proxy.to_string(): None   
```

**Comment by davidMcneil** [comment]
Resolved by #7532
